OSASINFRA-3280: openstack: document etcd on local disk (#7664)
openshift/installer: docs/user/openstack/etcd-ephemeral-disk.md

# Moving etcd to an ephemeral local disk

You can move etcd from a root volume (Cinder) to a dedicated ephemeral local disk to prevent or resolve performance issues.

## Prerequisites

* An OpenStack cloud with Cinder being functional and enough available storage (at least 75 GB) to accommodate 3 Root Volumes for the OpenShift control plane.

> * An OpenStack cloud whith Nova not using Ceph as a backend for ephemeral disks.

Reviewer: This should be moved to top as I think this is the number 1 requirement. Also, perhaps it is clearer if we don't use negative:
```suggestion
* An OpenStack cloud where Nova is configured to use local storage for ephemeral disks. The `libvirt.images_type` option in `nova.conf` must not be `rbd`.
```

* OpenShift will be deployed with IPI for now; UPI is not yet documented but technically possible.

> * The MachineConfigPool must match metadata.labels[machineconfiguration.openshift.io/role]. This applies to a controller, worker, or a custom pool.

Reviewer: ```suggestion
* The MachineConfigPool must match `metadata.labels[machineconfiguration.openshift.io/role]`. This applies to a controller, worker, or a custom pool.
```

There are a few more things I noticed in this PR that should be enclosed in backticks, like file names or path, or commands.

> * The control-plane machine’s auxiliary storage device, such as /dev/vdb, must match the vdb. Change this reference in all places in the file.

Reviewer 2: would it be worth to share commands on how to do that?

Author: Yes I sometimes think the same but I also find it painful to maintain the commands which often change over releases. I wonder the same about the amount of details sometimes.

I think the OCP docs should provide more details for the customers, but here in this repo I think it's a best-effort kind of documentation from what I've seen so far so in this case I'm fine if we don't provide that level of details.

Author: I'm in fact removing that line which is more confusing than helpful.

## Procedure

* Create a Nova flavor for the Control Plane which allows 10 GiB of Ephemeral Disk:

```bash

> openstack flavor create --ram 16384 --disk 25 --ephemeral 10 --vcpus 4 m1.xlarge

Reviewer: That is probably not enough disk space. We recommend at least 100GB disk space for the control plane nodes: https://docs.openshift.com/container-platform/4.14/installing/installing_openstack/installing-openstack-installer-custom.html#installation-osp-control-machines_installing-openstack-installer-custom

Also, the recommended memory and CPU depends on the expected cluster size: https://docs.openshift.com/container-platform/4.14/scalability_and_performance/recommended-performance-scale-practices/recommended-control-plane-practices.html#master-node-sizing_recommended-control-plane-practices

Author: I'll remove the RAM, disk and CPU numbers, the only thing of value here is the ephemeral flag.

```

* If you are deploying a new cluster, run openshift-install with the following parameters to create the manifests:

```bash

> openshift-install create manifests --dir=install_dir

Reviewer 2: Nice! I though moving etcd to local disk was only a day2 operation

Author: It will be a day 2 operation. Trying it as a day 1 was a mistake and it leads to unstable cluster right after a deployment. I suggest we don't recommend that **for now**.

```

> * If you are deploying a new cluster, edit the ControlPlaneMachineSet (CPMS) in `install-dir/openshift/99_openshift-machine-api_master-control-plane-machine-set.yaml` to add the additional block ephemeral device that will be used by etcd:

Reviewer: We can probably merge these steps into one. For example:

* edit the `ControlPlaneMachineSet` (CPMS) to add the additional block ephemeral device that will be used by etcd.
```yaml
[...]
providerSpec:
  additionalBlockDevices:
  - name: etcd
    sizeGiB: 10
    storage:
      type: Local
[...]
```
If you are deploying a new cluster, edit the configuration in `install-dir/openshift/99_openshift-machine-api_master-control-plane-machine-set.yaml` otherwise edit the CPMS on an existing cluster with `oc edit controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster`.

```yaml
[...]
providerSpec:
  additionalBlockDevices:
  - name: etcd
    sizeGiB: 10
    storage:
      type: Local

> [...]

Reviewer 2: What do you think about adding another example of `additionalBlockDevices` with volume type given this is a new API? Perhaps mention it in the customization docs to not mix with etcd things?

Author: I don't want to advertise that simply because:

1. Our CI doesn't and won't test it in the short term.
2. putting etcd on a volume is not recommended at the moment.

the control-plane machines.

* If you are deploying a new cluster, we can now deploy:

```bash
openshift-install create cluster --dir=install_dir
```

* In both scenarios, once the cluster is healthy, create a file named `etcd-replace.yaml` with this content:

```yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: master
  name: 98-var-lib-etcd
spec:
  config:
    ignition:
      version: 3.2.0
    systemd:
      units:
      - contents: |
          [Unit]
          Description=Mount /dev/vdb to /var/lib/etcd
          Before=local-fs.target
          Requires=systemd-mkfs@dev-vdb.service
          After=systemd-mkfs@dev-vdb.service var.mount

          [Mount]
          What=/dev/vdb
          Where=/var/lib/etcd
          Type=xfs
          Options=defaults,prjquota

          [Install]
          WantedBy=local-fs.target
        enabled: true
        name: var-lib-etcd.mount
```

Apply this file that will remove the logic for creating and syncing the device by entering the following command:

```bash
oc replace -f etcd-replace.yaml
```

> etcd is now running on the local disk.

Reviewer: How can we verify this is effectively the case?

# Moving etcd to an ephemeral local disk

You can move etcd from a root volume (Cinder) to a dedicated ephemeral local disk to prevent or resolve performance issues.

## Prerequisites

* This migration is currently tested and documented as a day 2 operation.
* An OpenStack cloud where Nova is configured to use local storage for ephemeral disks. The `libvirt.images_type` option in `nova.conf` must not be `rbd`.
* An OpenStack cloud with Cinder being functional and enough available storage to accommodate 3 Root Volumes for the OpenShift control plane.
* OpenShift will be deployed with IPI for now; UPI is not yet documented but technically possible.
* The control-plane machine’s auxiliary storage device, such as /dev/vdb, must match the vdb. Change this reference in all places in the file.

## Procedure

* Create a Nova flavor for the Control Plane which allows 10 GiB of Ephemeral Disk:

```bash
openstack flavor create --ephemeral 10 [...]
```

* We will deploy a cluster with Root Volumes for the Control Plane. Here is an example of `install-config.yaml`:

```yaml
[...]
controlPlane:
  name: master
  platform:
    openstack:
      type: ${CONTROL_PLANE_FLAVOR}
      rootVolume:

>         size: 25

Reviewer: 25 is not a recommended value, let's use something bigger

Author: doc says:

```
* `size` (required integer): Size of the root volume in GB. Must be set to at least 25.
```

And I'm happy to put more but then we need to be consistent in our doc and in the installer validations.

Reviewer: 25 is the minimum value and is fine for throwaway clusters like we do in CI. For production [we recommend at least 100](https://docs.openshift.com/container-platform/4.14/installing/installing_openstack/installing-openstack-installer-custom.html#installation-osp-control-machines_installing-openstack-installer-custom).

[The validations](https://github.com/openshift/installer/blob/a1ea1c8/pkg/asset/installconfig/openstack/validation/machinepool.go#L19-L20) reflect these values.

I've pushed https://github.com/openshift/installer/pull/7698 to clarify the docs.

        types:
        - ${CINDER_TYPE}
  replicas: 3
[...]
```

* Run openshift-install with the following parameters to create the cluster:

```bash
openshift-install create cluster --dir=install_dir
```

* Once the cluster has been deployed and is healthy, edit the ControlPlaneMachineSet (CPMS) to add the additional block ephemeral device that will be used by etcd:

```bash
oc patch ControlPlaneMachineSet/cluster -n openshift-machine-api --type json -p '[{"op": "add", "path": "/spec/template/machines_v1beta1_machine_openshift_io/spec/providerSpec/value/additionalBlockDevices", "value": [{"name": "etcd", "sizeGiB": 10, "storage": {"type": "Local"}}]}]'
```

> NOTE: Putting etcd on a block device of type Volume is not supported for performance reasons.

Reviewer: If you want the note to stand out, you can use [the github additions to markdown](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts):
```suggestion
> [!NOTE]
> Putting etcd on a block device of type Volume is not supported for performance reasons.
```

Now that raises the question of how etcd on an additional volume on ceph is different from keeping etcd on the root volume on ceph, and why it wouldn't be supported.

Author: I'll add a few words and use the new NOTE syntax.

oc wait --timeout=90m --for=jsonpath='{.spec.replicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
oc wait --timeout=90m --for=jsonpath='{.status.updatedReplicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
oc wait --timeout=90m --for=jsonpath='{.status.replicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
oc wait --timeout=90m --for=jsonpath='{.status.readyReplicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
oc wait clusteroperators --timeout=30m --all --for=condition=Progressing=false
```

* Check that we have 3 control plane machines, and that each machine has the additional block device:

```bash
cp_machines=$(oc get machines -n openshift-machine-api --selector='machine.openshift.io/cluster-api-machine-role=master' --no-headers -o custom-columns=NAME:.metadata.name)
if [[ $(echo "${cp_machines}" | wc -l) -ne 3 ]]; then
  exit 1
fi
for machine in ${cp_machines}; do
  if ! oc get machine -n openshift-machine-api "${machine}" -o jsonpath='{.spec.providerSpec.value.additionalBlockDevices}' | grep -q 'etcd'; then
  exit 1
  fi
done
```

> * 

Reviewer: Extra bullet point

```yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: master
  name: 98-var-lib-etcd
spec:
  config:
    ignition:
      version: 3.2.0
    systemd:
      units:
      - contents: |
          [Unit]
          Description=Mount /dev/vdb to /var/lib/etcd
          Before=local-fs.target
          Requires=systemd-mkfs@dev-vdb.service
          After=systemd-mkfs@dev-vdb.service var.mount

          [Mount]
          What=/dev/vdb
          Where=/var/lib/etcd
          Type=xfs
          Options=defaults,prjquota

          [Install]
          WantedBy=local-fs.target
        enabled: true
        name: var-lib-etcd.mount
```

Apply this file that will remove the logic for creating and syncing the device by entering the following command:

```bash
oc replace -f etcd-replace.yaml
```

* Again we need to wait for the cluster to be healthy. The same commands as above can be used to check that everything is healthy.

* Now etcd is stored on ephemeral local disk. This can be verified by connected to a master nodes with `oc debug node/<master-node-name>` and running the following commands:

```bash

> chroot /host

Reviewer: You can get the equivalent with a single liner:
```suggestion
oc debug node/<master-node-name> -- df -T /host/var/lib/etcd
```

Reviewer: This will give you the mount point, the file system type, and used space.
